File: src/main/java/com/tms/csp/ast/True.java
package com.tms.csp.ast;


import com.tms.csp.util.varSets.VarSet;

import java.math.BigInteger;

public class True extends Constant {

    private Exp neg;

    True(Space space, int expId) {
        super(space, expId);
    }

    @Override
    public boolean isConstantTrue() {
        return true;
    }

    @Override
    public boolean isSat() {
        return true;
    }

    @Override
    public BigInteger getSatCount() {
        System.err.println("True.getSatCount");
        return BigInteger.ONE;
    }

    @Override
    public int getCubeCount() {
        return 1;
    }

    @Override
    public void serialize(Ser a) {
        a.constantTrue();
    }

    final public Exp flip() {
        return getNeg();
    }

    @Override
    public boolean isPos() {
        return true;
    }

    @Override
    final public Exp getPos() {
        return this;
    }

    @Override
    final public Exp getNeg() {
        if (this.neg == null) {
            this.neg = _space.mkFalse();
        }
        return this.neg;
    }


    @Override
    public void serializeTinyDnnf(Ser a) {
        a.append(TRUE_CHAR);
    }

    @Override
    public void toXml(Ser a, int depth) {
        a.indent(depth);
        a.constantTrue();
    }

    public long satCountPL() {
        return 1;
    }

}

File: src/main/java/com/smartsoft/csp/ast/True.kt
package com.smartsoft.csp.ast


import com.smartsoft.csp.ast.PLConstants.TRUE_CHAR
import java.math.BigInteger

class True internal constructor(space: Space, expId: Int) : Constant(space, expId) {

    private var _neg: Exp? = null

    override val isConstantTrue: Boolean get() = true

    override val isSat: Boolean get() = true

    override val satCount: BigInteger get() = BigInteger.ONE

    override val cubeCount: Int get() = 1

    override val isPos: Boolean get() = true

    override val pos: Exp get() = this

    override fun serialize(a: Ser) {
        a.constantTrue()
    }

    override val flip: Exp get() = neg

    override val neg: Exp get() {
        if (this._neg == null) {
            this._neg = _space.mkFalse()
        }
        return this._neg!!
    }


    override fun serializeTinyDnnf(a: Ser) {
        a.append(TRUE_CHAR)
    }

    override fun toXml(a: Ser, depth: Int) {
        a.indent(depth)
        a.constantTrue()
    }

    override val satCountPL: Long get() = 1

}